src/Store/MenuStoreConnector.re

```diff
@@ -9,177 +9,270 @@ module Model = Oni_Model;
+              },
+              icon:
+                Oni_Model.FileExplorer.getFileIcon(
+                  languageInfo,
+                  iconTheme,
+                  path,
+                ),
+            })
+          | None => None
+          };
+        })
+    |> Array.of_seq;
+  };
+
+  let menuUpdater = (state: option(Quickmenu.t), action: Actions.t, buffers, languageInfo, iconTheme)
+    : (option(Quickmenu.t), Isolinear.Effect.t(Actions.t)) => {
     switch (action) {
-    | MenuSetLoading(isLoading, time) => (
-        {
-          ...state,
-          isLoading,
-          loadingAnimation: Animation.start(time, state.loadingAnimation),
-        },
-        Isolinear.Effect.none,
-      )
-    | MenuPosition(index) => (
-        {...state, selectedItem: index},
-        Isolinear.Effect.none,
-      )
-    | MenuPreviousItem => (
-        {
-          ...state,
-          selectedItem:
-            position(state.selectedItem, -1, filteredCommandsCount),
-        },
-        Isolinear.Effect.none,
-      )
-    | MenuNextItem => (
+    | MenuShow(CommandPalette) =>
+      (
+        Some{{
+          ...Quickmenu.defaults(CommandPalette),
+          source: Complete(Model.CommandPalette.commands)
+        }},
+        Isolinear.Effect.none
+      );
+
+    | MenuShow(Buffers) =>
+      (
+        Some{{
+          ...Quickmenu.defaults(Buffers),
+          source: Complete(makeBufferCommands(languageInfo, iconTheme, buffers))
+        }},
+        Isolinear.Effect.none
+      );
+
+    | MenuShow(WorkspaceFiles) =>
+      (
+        Some{{
+          ...Quickmenu.defaults(WorkspaceFiles),
+          source: Loading
+        }},
+        Isolinear.Effect.none
+      );
+
+    | MenuShow(Wildmenu(cmdType)) =>
+      (
+        Some{{
+          ...Quickmenu.defaults(Wildmenu(cmdType)),
+          prefix: Some(prefixFor(cmdType)),
+        }},
+        Isolinear.Effect.none
+      );
+
+    | MenuInput({ text, cursorPosition }) =>
+      (
+        Option.map(state => Quickmenu.{ ...state, text, cursorPosition }, state),
+        Isolinear.Effect.none
+      );
+
+    | MenuUpdateSource(source) =>
+      (
+        Option.map((state: Quickmenu.t) => {
+          let count = Quickmenu.getCount(source);
+          {...state, source, selected: min(count, state.selected)}
+        }, state),
+        Isolinear.Effect.none
+      );
+
+    | MenuFocus(index) => (
+      Option.map((state: Quickmenu.t) => {
+        let count = Quickmenu.getCount(state.source);
+
         {
           ...state,
-          selectedItem:
-            position(state.selectedItem, 1, filteredCommandsCount),
-        },
-        Isolinear.Effect.none,
-      )
-    | MenuSearch(query) => (
+          selected: max(0, min(count, index)) // TODO: Could use a clamp function
+        }
+      }, state),
+      Isolinear.Effect.none
+    )
+
+    | NotifyKeyPressed(_, "<UP>")
+    | MenuFocusPrevious => (
```

Comment:
This would be much less fragile if `NotifyKeyPressed` had carried a strongy typed key, but doing so would involve normalizing `keyEventListener` in `InputStoreHandler` on `keyEvent` instead of on a string.